add document for offline installation (#33)
chaos-mesh/website: docs/user_guides/offline_installation.md

---
id: offline_installation
title: Offline Installation
---

This document describes how to install Chaos Mesh in an offline environment. 

## Prerequisites

Before deploying Chaos Mesh, make sure the following items have been installed:

- Kubernetes version >= 1.12
- [RBAC](https://kubernetes.io/docs/admin/authorization/rbac) enabled (optional)
- Docker

## Prepare the installation file

To install Chaos Mesh offline,  you need to get the installation images via an internet connection. Take the steps below:

1. Specify the version you want to install:

```bash
export CHAOS_MESH_VERSION="v1.1.0"
```

> **Note:**
>
> It is recommended that you use a stable release, or you can set the version to `latest` if you want to experience the latest features that are under development.

2. Archive the docker images of Chaos Mesh:

> ```bash  #pull images of Chaos Mesh

Reviewer: Indent needed here and below

Author: done

docker save -o ./image-chaos-daemon pingcap/chaos-daemon:${CHAOS_MESH_VERSION}
docker save -o ./image-chaos-dashboard pingcap/chaos-dashboard:${CHAOS_MESH_VERSION}
```

3. Download the Chaos Mesh repository to your local:

```bash
wget "https://github.com/chaos-mesh/chaos-mesh/archive/${CHAOS_MESH_VERSION}.zip"
```

Or you can download the latest unstable version:

```bash
wget https://github.com/chaos-mesh/chaos-mesh/archive/master.zip
```

4. Copy the ./image-chaos-mesh`, `./image-chaos-daemon`, `./image-chaos-dashboard` and `{CHAOS_MESH_VERSION}.zip` into the offline environment.

## Install Chaos Mesh offline

Now that you already have the image and repo archive files in the offline environment, start installing Chaos Mesh.

1. Specify the version you are going to install in the offline environment:

   ```bash
   export CHAOS_MESH_VERSION="v1.1.0"
   ```

2. Load the image from the archive files:

   ```bash
   docker load -i ./image-chaos-mesh
   docker load -i ./image-chaos-daemon
   docker load -i ./image-chaos-dashboard
   ```

3. Push the Chaos Mesh images. You can choose to push them to Docker Registry or Docker Hub.

> - Push images to Docker Registry

Reviewer: ```suggestion
    - Push images to Docker Registry
```

Reviewer: More indent here.

>    Set the Docker Registry variable, for example:

Reviewer: ```suggestion
   a. Set the Docker Registry variable, for example:
```

Reviewer: Add sequence label for these steps.

Author: done

docker save -o ./image-chaos-mesh pingcap/chaos-mesh:${CHAOS_MESH_VERSION}
docker save -o ./image-chaos-daemon pingcap/chaos-daemon:${CHAOS_MESH_VERSION}
docker save -o ./image-chaos-dashboard pingcap/chaos-dashboard:${CHAOS_MESH_VERSION}
```

3. Download the Chaos Mesh repository to your local:

```bash
wget "https://github.com/chaos-mesh/chaos-mesh/archive/${CHAOS_MESH_VERSION}.zip"
```

Or you can download the latest unstable version:

```bash
wget https://github.com/chaos-mesh/chaos-mesh/archive/master.zip
```

4. Copy the ./image-chaos-mesh`, `./image-chaos-daemon`, `./image-chaos-dashboard` and `{CHAOS_MESH_VERSION}.zip` into the offline environment.

## Install Chaos Mesh offline

Now that you already have the image and repo archive files in the offline environment, start installing Chaos Mesh.

Specify the version you are going to install in the offline environment:

```bash #specify the version you are going to install
export CHAOS_MESH_VERSION="v1.1.0"
```

1. Load and push images

Load the image from archive files:

```bash #load images of Chaos Mesh
docker load -i ./image-chaos-mesh
docker load -i ./image-chaos-daemon
docker load -i ./image-chaos-dashboard
```

> You can choose to push these images to Docker Registry or Docker Hub.

Reviewer: ```suggestion
3. Push the Chaos Mesh images. You can choose to push them to Docker Registry or Docker Hub.
```

   ```bash #push tags of Chaos Mesh to Docker Registry
   docker push $CHAOS_MESH_IMAGE
   docker push $CHAOS_DAEMON_IMAGE
   docker push $CHAOS_DASHBOARD_IMAGE
   ```

   > **Note:**
   >
   > If the Docker Registry can only work locally, you need to load and push these images on each K8s node.

- Push images to Docker Hub

   Set the Docker Hub variable, for example:

   ```bash
   export DOCKER_HUB=hub
   ```

   Tag these images with `$DOCKER_REGISTRY`:

   ```bash #tag images of Chaos Mesh
   export CHAOS_MESH_IMAGE=$DOCKER_HUB/chaos-mesh:${CHAOS_MESH_VERSION}
   export CHAOS_DAEMON_IMAGE=$DOCKER_HUB/chaos-daemon:${CHAOS_MESH_VERSION}
   export CHAOS_DASHBOARD_IMAGE=$DOCKER_HUB/chaos-dashboard:${CHAOS_MESH_VERSION}
   docker image tag pingcap/chaos-mesh:${CHAOS_MESH_VERSION} $CHAOS_MESH_IMAGE
   docker image tag pingcap/chaos-daemon:${CHAOS_MESH_VERSION} $CHAOS_DAEMON_IMAGE
   docker image tag pingcap/chaos-dashboard:${CHAOS_MESH_VERSION} $CHAOS_DASHBOARD_IMAGE
   ```

   Then push these images to Docker Registry:

   ```bash #push tags of Chaos Mesh to Docker hub
   docker push $CHAOS_MESH_IMAGE
   docker push $CHAOS_DAEMON_IMAGE
   docker push $CHAOS_DASHBOARD_IMAGE
   ```

> 2. Installation

Reviewer: ```suggestion
2. Install Chaos Mesh offline with the following steps:
```

>    a. Unzip the repo archive files:

Reviewer: ```suggestion
   a. Unzip the repo archive files to a path:
```

   ```bash #unzip archive file of repo
   unzip {CHAOS_MESH_VERSION}.zip chaos-mesh
   ```

>    And then enter the directory.

Reviewer: Please keep all the step info in one place.

>    And then enter the directory.

Reviewer: ```suggestion
```

>    b. Create a namespace for install Chaos Mesh:

Reviewer: ```suggestion
   b. Create a namespace for installing Chaos Mesh:
```

>    ```bash #create a namespace for install Chaos Mesh

Reviewer: ```suggestion
   ```bash  
```

---
id: offline_installation
title: Offline Installation
---

> This document describes how to install Chaos Mesh under an offline environment, which means can not connect to the internet.

Author: The production environment of some state-owned enterprises has strict network requirements and maybe cannot connect to external networks

## Prerequisites

Before deploying Chaos Mesh, make sure the following items have been installed:

- Kubernetes version >= 1.12
- [RBAC](https://kubernetes.io/docs/admin/authorization/rbac) enabled (optional)
- Docker

## Prepare the installation file

You need to prepare the installation file with a computer connected to the internet. 

You can specify which version you want to install:

```bash
export CHAOS_MESH_VERSION="v1.1.0"
```

> It is recommended that you use a stable release, or you can set the version to `latest` if you want to experience the latest features that are under development. 

Author: done

### Archive docker images

First, you need to pull images of Chaos Mesh:

```bash
docker pull pingcap/chaos-mesh:${CHAOS_MESH_VERSION}
docker pull pingcap/chaos-daemon:${CHAOS_MESH_VERSION}
docker pull pingcap/chaos-dashboard:${CHAOS_MESH_VERSION}
```

> Then save these images to tar archives:

Author: done

```bash
docker save -o ./image-chaos-mesh pingcap/chaos-mesh:${CHAOS_MESH_VERSION}
docker save -o ./image-chaos-daemon pingcap/chaos-daemon:${CHAOS_MESH_VERSION}
docker save -o ./image-chaos-dashboard pingcap/chaos-dashboard:${CHAOS_MESH_VERSION}
```

### Download Chaos Mesh repo

Download Chaos Mesh repo with release version:

```bash
wget "https://github.com/chaos-mesh/chaos-mesh/archive/${CHAOS_MESH_VERSION}.zip"
```

> Or you can download the latest unstable version:

Author: For the latest images, the image name is `...:latest`, but for the repo, it is `.../master.zip`, so I description it again.

> Or you can download the latest unstable version:

Author: We need them all, these two steps download different files.

```bash
wget https://github.com/chaos-mesh/chaos-mesh/archive/master.zip
```

### Copy image and repo archive files

Copy the file `./image-chaos-mesh`, `./image-chaos-daemon`, `./image-chaos-dashboard` and `{CHAOS_MESH_VERSION}.zip` into the offline environment.

## Install Chaos Mesh

Now you already have the image and repo archive files in the offline environment, you can use them to install Chaos Mesh.

> Specify the version you are going to install in the offline environment:

Author: The first one is executed in the computer which can connect to the network, this one is executed in the offline environment.

> Specify the version you are going to install in the offline environment:

Reviewer: Users need to connect to the network to prepare the installation files, with a specified version - this means the version is already specified for the installation, right?

> Specify the version you are going to install in the offline environment:

Author: Yes, but need to execute twice

> Specify the version you are going to install in the offline environment:

Reviewer: Maybe something to consider for improvement - this sounds unnecessary
